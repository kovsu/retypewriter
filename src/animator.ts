import type { Patch } from '.'

export function* createAnimator(input: string, patches: Patch[]) {
  let output = input
  let cursor = 0

  for (const patch of patches) {
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = output.slice(0, patch.from)
      const tail = output.slice(patch.from)
      let selection = ''
      for (const char of patch.text) {
        selection += char
        yield {
          cursor: cursor + selection.length,
          output: head + selection + tail,
        }
      }
      output = head + selection + tail
    }

    if (patch.type === 'removal') {
      for (let i = patch.length; i >= 0; i--) {
        yield {
          cursor: cursor - i,
          output: output.slice(0, patch.from - patch.length + i) + output.slice(patch.from),
        }
      }
      output = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
    }
  }

  return {
    output,
  }
}
