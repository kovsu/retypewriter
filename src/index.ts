import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface RemovalPatch {
  type: 'removal'
  from: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export function calculatePath(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  let index = 0
  for (const diff of diffs) {
    if (diff[0] === 0) {
      index += diff[1].length
      continue
    }

    if (diff[0] === -1) {
      patches.push({
        type: 'removal',
        from: index + diff[1].length,
        length: diff[1].length,
      })
    }

    if (diff[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: diff[1],
      })
      index += diff[1].length
    }
  }
  return patches
}

export function applyPatches(input: string, patches: Patch[]) {
  let output = input

  for (const patch of patches) {
    if (patch.type === 'insert') {
      output
        = output.slice(0, patch.from) + patch.text + output.slice(patch.from)
    }
    if (patch.type === 'removal') {
      output
        = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
    }
  }

  return { output }
}
