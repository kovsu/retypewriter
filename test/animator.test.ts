import { describe, expect, it } from 'vitest'
import { calculatePath, diff } from '../src'
import { craeteAnimator } from '../src/animator'

const input = `
  describe('should', () => {
    it('exported', () => {
      expect(one).toEqual(1)
    })
  })
`

const output = `
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual(2)
      expect(one).toEqual(1)
    })
  })
`

describe('should', () => {
  it('exported', () => {
    const delta = diff(input, output)
    const patches = calculatePath(delta)

    const animator = craeteAnimator(input, patches)
    expect([...animator]).toMatchSnapshot('animator')
  })
})
