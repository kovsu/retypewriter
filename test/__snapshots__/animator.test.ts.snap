// Vitest Snapshot v1

exports[`should > exported > animator 1`] = `
[
  {
    "cursor": -8,
    "output": "
  describe('should', () => {
    it('exported', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -7,
    "output": "
  describe('should', () => {
    it('exporte', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -6,
    "output": "
  describe('should', () => {
    it('export', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -5,
    "output": "
  describe('should', () => {
    it('expor', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -4,
    "output": "
  describe('should', () => {
    it('expo', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -3,
    "output": "
  describe('should', () => {
    it('exp', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -2,
    "output": "
  describe('should', () => {
    it('ex', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": -1,
    "output": "
  describe('should', () => {
    it('e', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 0,
    "output": "
  describe('should', () => {
    it('', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 39,
    "output": "
  describe('should', () => {
    it('h', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 40,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 52,
    "output": "
  describe('should', () => {
    it('hi', () => {
       expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 53,
    "output": "
  describe('should', () => {
    it('hi', () => {
        expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 54,
    "output": "
  describe('should', () => {
    it('hi', () => {
         expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 55,
    "output": "
  describe('should', () => {
    it('hi', () => {
          expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 56,
    "output": "
  describe('should', () => {
    it('hi', () => {
           expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 57,
    "output": "
  describe('should', () => {
    it('hi', () => {
            expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 58,
    "output": "
  describe('should', () => {
    it('hi', () => {
      e      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 59,
    "output": "
  describe('should', () => {
    it('hi', () => {
      ex      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 60,
    "output": "
  describe('should', () => {
    it('hi', () => {
      exp      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 61,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expe      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 62,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expec      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 63,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 64,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 65,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(t      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 66,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(tw      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 67,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 68,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two)      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 69,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 70,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).t      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 71,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).to      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 72,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toE      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 73,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEq      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 74,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqu      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 75,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqua      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 76,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 77,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual(      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 78,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual(2      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 79,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual(2)      expect(one).toEqual(1)
    })
  })
",
  },
  {
    "cursor": 80,
    "output": "
  describe('should', () => {
    it('hi', () => {
      expect(two).toEqual(2)
      expect(one).toEqual(1)
    })
  })
",
  },
]
`;
